File: edgequake/sdks/typescript/src/resources/tasks.ts
/**
 * Tasks resource — track async task status.
 *
 * @module resources/tasks
 * @see edgequake/crates/edgequake-api/src/handlers/tasks.rs
 */

import type {
  ListTasksQuery,
  TaskListResponse,
  TaskStatus,
} from "../types/tasks.js";
import { Resource } from "./base.js";

export class TasksResource extends Resource {
  /** Get task status by track ID. */
  async get(trackId: string): Promise<TaskStatus> {
    return this._get(`/api/v1/tasks/${trackId}`);
  }

  /** List tasks with optional filters and pagination. */
  async list(query?: ListTasksQuery): Promise<TaskListResponse> {
    const params = new URLSearchParams();
    if (query?.tenant_id) params.set("tenant_id", query.tenant_id);
    if (query?.workspace_id) params.set("workspace_id", query.workspace_id);
    if (query?.status) params.set("status", query.status);
    if (query?.task_type) params.set("task_type", query.task_type);
    if (query?.page !== undefined) params.set("page", String(query.page));
    if (query?.page_size !== undefined)
      params.set("page_size", String(query.page_size));
    if (query?.sort) params.set("sort", query.sort);
    if (query?.order) params.set("order", query.order);
    const qs = params.toString();
    return this._get(`/api/v1/tasks${qs ? `?${qs}` : ""}`);
  }

  /** Cancel a running task. */
  async cancel(trackId: string): Promise<void> {
    await this._post(`/api/v1/tasks/${trackId}/cancel`);
  }

  /** Retry a failed task. */
  async retry(trackId: string): Promise<TaskStatus> {
    return this._post(`/api/v1/tasks/${trackId}/retry`);
  }
}

File: edgequake/sdks/rust/src/resources/tasks.rs
//! Tasks resource.

use crate::client::EdgeQuakeClient;
use crate::error::Result;
use crate::types::operations::*;

pub struct TasksResource<'a> {
    pub(crate) client: &'a EdgeQuakeClient,
}

impl<'a> TasksResource<'a> {
    /// `GET /api/v1/tasks`
    pub async fn list(&self) -> Result<TaskListResponse> {
        self.client.get("/api/v1/tasks").await
    }

    /// `GET /api/v1/tasks/{track_id}`
    pub async fn get(&self, track_id: &str) -> Result<TaskInfo> {
        self.client
            .get(&format!("/api/v1/tasks/{track_id}"))
            .await
    }

    /// `POST /api/v1/tasks/{track_id}/cancel`
    pub async fn cancel(&self, track_id: &str) -> Result<()> {
        self.client
            .post_no_content::<()>(
                &format!("/api/v1/tasks/{track_id}/cancel"),
                None,
            )
            .await
    }

    /// `POST /api/v1/tasks/{track_id}/retry` — Retry a failed task.
    pub async fn retry(&self, track_id: &str) -> Result<serde_json::Value> {
        self.client
            .post::<(), serde_json::Value>(
                &format!("/api/v1/tasks/{track_id}/retry"),
                None,
            )
            .await
    }
}
